File: ios/CoachCore/Sources/CoachCore/Classifier.swift
// The trained exercise classifier — the SAME gated, versioned TinyMLP the
// desktop trains (pose_coach.py --train-classifier), exported as portable
// JSON (--export-model) and run here without any ML framework: ~1.5k
// parameters, a matrix multiply, a ReLU and a softmax. Parity with the
// Python engine is pinned by data/parity_fixtures.json (window_feature and
// mlp sections).

import Foundation

/// Fixed-size feature vector over a window of frames — port of the
/// desktop's window_features: per-channel mean/std/min/max (population
/// std!) plus torso-normalized shoulder & wrist travel.
public enum WindowFeaturesML {
    public static let ndim = 38

    static func channels(_ f: FrameFeatures) -> [Double] {
        [f.trunk, f.knee, f.elbow, f.hip, f.shoY, f.wriY, f.torso,
         f.overhead ? 1.0 : 0.0, f.kneeSplit]
    }

    private static let torsoIdx = 6
    private static let shoYIdx = 4
    private static let wriYIdx = 5

    public static func of(_ frames: [FrameFeatures]) -> [Double] {
        precondition(!frames.isEmpty, "window features need >= 1 frame")
        let rows = frames.map { channels($0) }
        let n = Double(rows.count)
        let nc = rows[0].count
        var mean = [Double](repeating: 0, count: nc)
        var mn = [Double](repeating: .infinity, count: nc)
        var mx = [Double](repeating: -.infinity, count: nc)
        for r in rows {
            for c in 0..<nc {
                mean[c] += r[c]
                mn[c] = Swift.min(mn[c], r[c])
                mx[c] = Swift.max(mx[c], r[c])
            }
        }
        for c in 0..<nc { mean[c] /= n }
        var std = [Double](repeating: 0, count: nc)
        for r in rows {
            for c in 0..<nc {
                let d = r[c] - mean[c]
                std[c] += d * d
            }
        }
        for c in 0..<nc { std[c] = (std[c] / n).squareRoot() }
        let torso = Swift.max(mean[torsoIdx], 1e-3)
        var out = [Double]()
        out.reserveCapacity(ndim)
        out.append(contentsOf: mean)
        out.append(contentsOf: std)
        out.append(contentsOf: mn)
        out.append(contentsOf: mx)
        out.append((mx[shoYIdx] - mn[shoYIdx]) / torso)
        out.append((mx[wriYIdx] - mn[wriYIdx]) / torso)
        return out
    }
}

/// Two-layer MLP inference: (x-mu)/sd -> ReLU hidden -> softmax classes.
public final class TinyMLP {
    public let classes: [String]
    public let minProba: Double
    public let modelVersion: String
    let w1: [[Double]]      // [ndim][hidden]
    let b1: [Double]
    let w2: [[Double]]      // [hidden][classes]
    let b2: [Double]
    let mu: [Double]
    let sd: [Double]

    public init(classes: [String], minProba: Double,
                w1: [[Double]], b1: [Double],
                w2: [[Double]], b2: [Double],
                mu: [Double], sd: [Double],
                modelVersion: String = "unknown") {
        self.classes = classes
        self.minProba = minProba
        self.w1 = w1
        self.b1 = b1
        self.w2 = w2
        self.b2 = b2
        self.mu = mu
        self.sd = sd
        self.modelVersion = modelVersion
    }

    /// Load a pose_coach.py --export-model file. Returns nil on any
    /// malformed input — a broken model must never crash a workout.
    public static func load(json data: Data) -> TinyMLP? {
        guard let root = (try? JSONSerialization.jsonObject(with: data))
                as? [String: Any],
              let classes = root["classes"] as? [String],
              let w1 = matrix(root["W1"]), let b1 = vector(root["b1"]),
              let w2 = matrix(root["W2"]), let b2 = vector(root["b2"]),
              let mu = vector(root["mu"]), let sd = vector(root["sd"])
        else { return nil }
        let minProba = (root["min_proba"] as? NSNumber)?.doubleValue ?? 0.75
        let version = ((root["manifest"] as? [String: Any])?["model_version"]
                       as? String) ?? "unknown"
        return TinyMLP(classes: classes, minProba: minProba,
                       w1: w1, b1: b1, w2: w2, b2: b2, mu: mu, sd: sd,
                       modelVersion: version)
    }

    public static func load(url: URL) -> TinyMLP? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return load(json: data)
    }

    private static func vector(_ any: Any?) -> [Double]? {
        (any as? [Any])?.compactMap { ($0 as? NSNumber)?.doubleValue }
    }

    private static func matrix(_ any: Any?) -> [[Double]]? {
        guard let rows = any as? [Any] else { return nil }
        var out: [[Double]] = []
        for row in rows {
            guard let v = vector(row) else { return nil }
            out.append(v)
        }
        return out
    }

    public func predict(_ x: [Double]) -> [Double] {
        var xn = [Double](repeating: 0, count: x.count)
        for i in 0..<x.count { xn[i] = (x[i] - mu[i]) / sd[i] }
        var h = [Double](repeating: 0, count: b1.count)
        for j in 0..<b1.count {
            var acc = b1[j]
            for i in 0..<xn.count { acc += xn[i] * w1[i][j] }
            h[j] = Swift.max(0.0, acc)
        }
        var z = [Double](repeating: 0, count: b2.count)
        for k in 0..<b2.count {
            var acc = b2[k]
            for j in 0..<h.count { acc += h[j] * w2[j][k] }
            z[k] = acc
        }
        let zmax = z.max() ?? 0
        let e = z.map { Foundation.exp($0 - zmax) }
        let tot = e.reduce(0, +)
        return e.map { $0 / tot }
    }
}

/// AutoDetector with the rule-based vote swapped for the trained MLP —
/// same sliding window, vote cadence, and 3-agreeing-votes lock-in as the
/// desktop.
public final class MLDetector: AutoDetector {
    let model: TinyMLP

    public init(model: TinyMLP) {
        self.model = model
        super.init()
    }

    override func classify() -> String? {
        let p = model.predict(WindowFeaturesML.of(buf.map { $0.1 }))
        guard let maxP = p.max(), let ci = p.firstIndex(of: maxP),
              maxP >= model.minProba else { return nil }
        return model.classes[ci]
    }
}

File: android/app/src/main/java/com/fekitech/gymcoach/core/Classifier.kt
// The trained exercise classifier — the SAME gated, versioned TinyMLP the
// desktop trains (pose_coach.py --train-classifier), exported as portable
// JSON (--export-model) and run here without any ML framework: ~1.5k
// parameters, a matrix multiply, a ReLU and a softmax. Parity with the
// Python engine is pinned by data/parity_fixtures.json (window_feature and
// mlp sections).
package com.fekitech.gymcoach.core

import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

/**
 * Fixed-size feature vector over a window of frames — port of the desktop's
 * window_features: per-channel mean/std/min/max (population std!) plus
 * torso-normalized shoulder & wrist travel. Channel order = FEAT_KEYS.
 */
object WindowFeatures {
    const val NDIM = 38

    private fun channels(f: FrameFeatures): DoubleArray = doubleArrayOf(
        f.trunk, f.knee, f.elbow, f.hip, f.shoY, f.wriY, f.torso,
        if (f.overhead) 1.0 else 0.0, f.kneeSplit)

    private const val TORSO = 6      // index of "torso" in FEAT_KEYS order
    private const val SHO_Y = 4
    private const val WRI_Y = 5

    fun of(frames: List<FrameFeatures>): DoubleArray {
        require(frames.isNotEmpty()) { "window_features needs >= 1 frame" }
        val rows = frames.map { channels(it) }
        val n = rows.size
        val nc = rows[0].size
        val mean = DoubleArray(nc)
        val mn = DoubleArray(nc) { Double.POSITIVE_INFINITY }
        val mx = DoubleArray(nc) { Double.NEGATIVE_INFINITY }
        for (r in rows) for (c in 0 until nc) {
            mean[c] += r[c]
            mn[c] = min(mn[c], r[c])
            mx[c] = max(mx[c], r[c])
        }
        for (c in 0 until nc) mean[c] /= n
        val std = DoubleArray(nc)
        for (r in rows) for (c in 0 until nc) {
            val d = r[c] - mean[c]
            std[c] += d * d
        }
        for (c in 0 until nc) std[c] = kotlin.math.sqrt(std[c] / n)
        val torso = max(mean[TORSO], 1e-3)
        val out = DoubleArray(NDIM)
        for (c in 0 until nc) {
            out[c] = mean[c]
            out[nc + c] = std[c]
            out[2 * nc + c] = mn[c]
            out[3 * nc + c] = mx[c]
        }
        out[4 * nc] = (mx[SHO_Y] - mn[SHO_Y]) / torso
        out[4 * nc + 1] = (mx[WRI_Y] - mn[WRI_Y]) / torso
        return out
    }
}

/** Two-layer MLP inference: (x-mu)/sd -> ReLU hidden -> softmax classes. */
class TinyMlp(
    val classes: List<String>,
    val minProba: Double,
    private val w1: Array<DoubleArray>,   // [NDIM][hidden]
    private val b1: DoubleArray,
    private val w2: Array<DoubleArray>,   // [hidden][classes]
    private val b2: DoubleArray,
    private val mu: DoubleArray,
    private val sd: DoubleArray,
    val modelVersion: String = "unknown",
) {
    fun predict(x: DoubleArray): DoubleArray {
        val xn = DoubleArray(x.size) { (x[it] - mu[it]) / sd[it] }
        val h = DoubleArray(b1.size) { j ->
            var acc = b1[j]
            for (i in xn.indices) acc += xn[i] * w1[i][j]
            max(0.0, acc)
        }
        val z = DoubleArray(b2.size) { k ->
            var acc = b2[k]
            for (j in h.indices) acc += h[j] * w2[j][k]
            acc
        }
        val zmax = z.max()
        var tot = 0.0
        val e = DoubleArray(z.size) { exp(z[it] - zmax).also { v -> tot += v } }
        return DoubleArray(e.size) { e[it] / tot }
    }

}

/**
 * AutoDetector with the rule-based vote swapped for the trained MLP — same
 * sliding window, vote cadence, and 3-agreeing-votes lock-in as the desktop.
 */
class MlDetector(private val model: TinyMlp) : AutoDetector() {
    override fun classify(): String? {
        val p = model.predict(WindowFeatures.of(windowFrames()))
        var ci = 0
        for (i in p.indices) if (p[i] > p[ci]) ci = i
        return if (p[ci] >= model.minProba) model.classes[ci] else null
    }
}
